# Basic stuff

At this first session we will discuss the following:
<ol type="A">
  <li>Numbers</li>
  <li>Strings</li>
  <li>Lists</li>
  <li>Dictionaries</li>
  <li>Tuples</li>
  <li>Sets</li>
  <li>Booleans</li>
</ol>

<hr>

## A. Numbers

How Python treat Numbers.

<ul style="list-style-type:none">
  <li>A.1 Types</li>
  <li>A.2 Arithmetic</li>
  <li>A.3 Variable Assignment</li>
</ul>

### A.1 Types

We'll focus on **integers** and **floating** point numbers, but there are other "types".
<br> **Integers**  are whole numbers, positive or negative: 1 and -8.
<br>**Floating** point numbers in Python are easy to see due to the decimal point or use the exponential (e) symbol. <code>1.0   -8.0   2.1  2e3 </code>

### A.2 Arithmetic

In [ ]:
# Addition
2+1

3

In [2]:
# Subtraction
2-1

1

In [3]:
# Multiplication
2*2

4

In [4]:
# Division
3/2

1.5

In [5]:
# Floor Division - truncate the decimal without rounding
7//4

1

In [6]:
# Modulo - remainder after division
7%4

3

In [7]:
# Powers
2**3

8

In [8]:
# Roots!
4**0.5

2.0

In [9]:
# Order of Operations
2 + 10 * 10 + 3

105

In [10]:
# Can use parentheses to specify orders
(2+10) * (10+3)

156

### A.3 Variable Assignment

In [11]:
# Let's create an object called "a" and assign it the number 5
a = 5

In [12]:
# Adding the objects
a+a

10

In [13]:
# Reassignment
a = 10

In [14]:
# Check
a

10

In [15]:
# Use A to redefine A
a = a + a

In [16]:
# Check 
a

20

Variable names have some rules and best practices...
<ol>
  <li>Names cannot start with a number</li>
  <li>No whitespaces - use <code>_</code></li>
  <li>Can't use any of these symbols:<code>'",<>/?|\()!@#$%^&*~-+</code></li>
  <li>It's considered best practice (PEP8) that names are lowercase</li>
  <li>Avoid using the characters 'l' (lowercase letter el), 'O' (uppercase letter oh), or 'I' (uppercase letter eye) as single character variable names.</li>
  <li>Avoid using words that have special meaning in Python like "list" and "str"</li>
</ol>

In [17]:
# Name write to help you keep track of things
my_income = 100

tax_rate = 0.1

my_taxes = my_income*tax_rate

In [18]:
# If you want to see your taxes it is easier to remember
my_taxes

10.0

<hr>
## B. Strings

Record text information. Strings in Python are actually a *sequence*, which basically means Python keeps track of every element in the string as a sequence. What that means? That you can use the indexes of the string to operate on it!

<ul style="list-style-type:none">
  <li>B.1 Creating</li>
  <li>B.2 Printing</li>
  <li>B.3 Indexing & Slicing</li>
  <li>B.4 String Properties</li>
  <li>B.5 Built-in Methods</li>
  <li>B.6 Print Format</li>
</ul>

### B.1 Creating

To create a string in Python you need to use either single quotes or double quotes.


In [19]:
# Single word
'KLM'

'KLM'

In [20]:
# Sentence 
'I just started to work here'

'I just started to work here'

In [21]:
# We can also use double quote
"I just started to work here"

'I just started to work here'

In [22]:
# Be careful with quotes!
'I'm at KLM'

SyntaxError: invalid syntax (<ipython-input-22-dd3f21a787ff>, line 2)

In [23]:
"I'm at KLM"

"I'm at KLM"

In [24]:
'I\'m at KLM'

"I'm at KLM"

### B.2 Printing
Using a notebook with just a string in a cell will automatically output strings, but the correct way to display strings in your output is by using a print function.

In [25]:
# Note that we can't output multiple strings this way
'This won\'t print'
'This will print'

'This will print'

In [26]:
print('Hello KLM')
print('I\'m from Brazil')
print('Use \n to print a new line')
print('\n')
print('New line in between!')

Hello KLM
I'm from Brazil
Use 
 to print a new line


New line in between!


In [27]:
# include space and punctuation
print(len('Hello KLM'))
len(' Hello, KLM!')

9


12

### B.3 Indexing & Slicing

Since strings are sequences we can use indexes to call parts of the sequence.

In Python, we use brackets <code>[]</code> after an object to call its index. Take note that indexing starts at 0 for Python.

In [28]:
s = 'Hello KLM'
s[0]

'H'

In [29]:
s[1]

'e'

In [30]:
# Grab everything past the first term all the way to the length of s which is len(s)
s[1:]

'ello KLM'

In [31]:
# Grab everything UP TO the 4th index
s[:4]

'Hell'

In [32]:
# Last letter (one index behind 0 so it loops back around)
s[-1]

'M'

In [33]:
# Grab everything but the last letter
s[:-1]

'Hello KL'

In [34]:
# Grab everything, but go in step sizes of 2
s[::2]

'HloKM'

In [35]:
# We can use this to print a string backwards
s[::-1]

'MLK olleH'

### B.4 Properties

Strings are **IMMUTABLE**.

In [36]:
s

'Hello KLM'

In [37]:
s[0] = 'x'

TypeError: 'str' object does not support item assignment

Notice how the error tells us directly what we can't do, change the item assignment!

Something we can do is **CONCATENATE** strings!

In [38]:
s + 'concatenate me'

'Hello KLMconcatenate me'

We can use the multiplication symbol <em><code>*</code></em> to create repetition!

### B.5 Built-in methods

Objects in Python usually have built-in methods. These methods are functions inside the object that can perform actions or commands on the object itself.
<br>We call methods with a period and then the method name. Methods are in the form:

`object.method(parameters)`

Where parameters are extra arguments we can pass into the method.

In [39]:
s.upper()

'HELLO KLM'

In [40]:
s.lower()

'hello klm'

In [41]:
# Split a string by blank space (this is the default)
s.split()

['Hello', 'KLM']

In [42]:
# Split by a specific element (doesn't include the element that was split on)
s.split('l')

['He', '', 'o KLM']

### B.6 Print Formatting

We can use the .format() method to add formated objects to printed string statements. 

In [43]:
# Old way
'%s %s' % ('one', 'two')

'one two'

In [44]:
# New way
'{} {}'.format('one','two')

'one two'

In [45]:
# placeholder can be confusing
'{1} {0}'.format('one','two')

'two one'

In [46]:
# Padding and Aligning

# 10 characters - align right
'{:>10}'.format('KLM')

'       KLM'

In [47]:
# 10 characters - align left
'{:10}'.format('KLM')

'KLM       '

In [48]:
# 10 characters - align right choosing the character
'{:_>10}'.format('KLM')

'_______KLM'

In [49]:
# 10 characters - center align
'{:^10}'.format('KLM')

'   KLM    '

In [50]:
# truncate & padding
'{:^10.3}'.format('KLM')

'   KLM    '

In [51]:
# Integers
print('{:d}'.format(37))
'{:d}'.format(3.14159)

37


ValueError: Unknown format code 'd' for object of type 'float'

In [52]:
'{:d}'.format(int(3.14159))

'3'

In [53]:
'{:f}'.format(3.14159)

'3.141590'

In [54]:
# pading integer number
'{:^4d}'.format(37)

' 37 '

In [55]:
# pading floating number
## '{:$atLeast.$afterDecimal$f}'.format(3.14159)

print('{:1.1f}'.format(3.14159))
print('{:5.1f}'.format(3.14159))
'{:5.5f}'.format(3.14159)

3.1
  3.1


'3.14159'

In [56]:
# Named placeholder
data = {'first': 'Hold', 'last':'the door!'}
'{first} {last}'.format(**data)

'Hold the door!'

In [57]:
# Datetime
from datetime import datetime
'{:%Y-%m-%d %H:%M}'.format(datetime(2018, 3, 19, 17, 23))

'2018-03-19 17:23'

In [58]:
# parametric
'{:{align}{width}}'.format('KLM', align='^', width='10')

'   KLM    '

<hr>
## C. Lists

Earlier when discussing strings we introduced the concept of a *sequence* in Python. Lists can be thought of the most general version of a *sequence* in Python. Unlike strings, they are mutable, meaning the elements inside a list can be changed!

<ul style="list-style-type:none">
  <li>C.1 Creating</li>
  <li>C.2 Indexing & Slicing</li>
  <li>C.3 Built-in Methods</li>
  <li>C.4 Nesting</li>
</ul>

### C.1 Creating


In [59]:
# Assign a list to an variable named my_list
my_list = [1,2,3]

In [60]:
my_mixed_list = ['A string',78, 3.14, 'o']

In [61]:
len(my_mixed_list)

4

### C.2 Indexing & Slicing

It works the same as the **String**.

In [62]:
my_mixed_list[0]

'A string'

In [63]:
my_mixed_list[1:3]

[78, 3.14]

In [64]:
# You can also concatenate
my_mixed_list + ["concatenate again"]

['A string', 78, 3.14, 'o', 'concatenate again']

In [65]:
# You can also repeat lists
my_mixed_list * 3

['A string',
 78,
 3.14,
 'o',
 'A string',
 78,
 3.14,
 'o',
 'A string',
 78,
 3.14,
 'o']

In [66]:
# Play around with double indexes
my_mixed_list[0][3]

't'

In [67]:
# Out of range
my_mixed_list[6]

IndexError: list index out of range

### C.3 Bult-in Methods

Do lists remind you of <code>array</code>? It surely is close to it, but Lists in Python tend to be more flexible than arrays in other languages. Why?
<br> Because they have no fixed size (meaning we don't have to specify how big a list will be), and they have no fixed type constraint (like we've seen above).

In [68]:
# Create a new list
list1 = [1,2,3]
# Append
list1.append('A string in an integer array? Outrageous!')
list1

[1, 2, 3, 'A string in an integer array? Outrageous!']

In [69]:
# Pop off the 0 indexed item
list1.pop(0)
list1

[2, 3, 'A string in an integer array? Outrageous!']

In [70]:
list1

[2, 3, 'A string in an integer array? Outrageous!']

In [70]:
list1.reverse()
list1

['A string in an integer array? Outrageous!', 3, 2]

In [71]:
# Use sort to sort the list - alphabetical and ascending
list1.sort()
list1

TypeError: '<' not supported between instances of 'int' and 'str'

In [72]:
only_string_list = ['a','e','x','b','c']
only_string_list.sort()
only_string_list

['a', 'b', 'c', 'e', 'x']

### C.4 Nesting Lists
A great feature of of Python data structures is that they support *nesting*. This means we can have data structures within data structures. For example: A list inside a list.

In [73]:
lst_1=[1,3,5]
lst_2=[2,4,6]
lst_3=[7,8,9]

# Form a matrix
list_of_lists = [lst_1,lst_2,lst_3]
list_of_lists

[[1, 3, 5], [2, 4, 6], [7, 8, 9]]

In [74]:
list_of_lists[0]

[1, 3, 5]

In [75]:
list_of_lists[0][1]

3

<hr>

## D. Dictionaries

These are **mappings** - similar to hash tables in other languages. Mappings are a collection of objects that are stored by a *key*, unlike a sequence that stored objects by their relative position. This is an important distinction, since mappings won't retain order since they have objects defined by a key.

<ul style="list-style-type:none">
  <li>D.1 Creating</li>
  <li>D.2 Nesting</li>
  <li>D.3 Built-in Methods</li>
</ul>

### D.1 Creating 
A Python dictionary consists of a key and then an associated value. That value can be almost any Python object.


In [76]:
# {} and ":"
my_dict = {'flight_number':889,'airplane_type':'Embraer'}

In [77]:
# Call values by their key
my_dict['flight_number']

889

In [78]:
# Very flexible
my_dict = {'flight_number':889, 'seat_row':[39,39,37],'attendants':['John','Mary','Marloes']}

In [79]:
# Let's call items from the dictionary
my_dict['attendants']

['John', 'Mary', 'Marloes']

In [80]:
my_dict['attendants'][0]

'John'

In [81]:
my_dict['attendants'][0].upper()

'JOHN'

In [82]:
## It is mutable!
my_dict['seat_row'][0] = my_dict['seat_row'][0] + 1
my_dict['seat_row']

[40, 39, 37]

In [83]:
# Create a new key through assignment
my_dict['airplane_type'] = 'Embraer'
my_dict

{'airplane_type': 'Embraer',
 'attendants': ['John', 'Mary', 'Marloes'],
 'flight_number': 889,
 'seat_row': [40, 39, 37]}

### D.2 Nesting

In [84]:
# Dictionary nested inside a dictionary nested inside a dictionary
d = {'key1':{'nestkey':{'subnestkey':'value'}}}
d['key1']['nestkey']['subnestkey']

'value'

### D.3 Built-in Methods

In [85]:
my_dict.keys()

dict_keys(['flight_number', 'seat_row', 'attendants', 'airplane_type'])

In [86]:
my_dict.values()

dict_values([889, [40, 39, 37], ['John', 'Mary', 'Marloes'], 'Embraer'])

In [87]:
my_dict.items()

dict_items([('flight_number', 889), ('seat_row', [40, 39, 37]), ('attendants', ['John', 'Mary', 'Marloes']), ('airplane_type', 'Embraer')])

<hr>

## E. Tuples

They are like **immutable** lists.

<ul style="list-style-type:none">
  <li>E.1 Creating</li>
  <li>E.2 Built-in Methods</li>
  <li>E.3 Usage</li>
</ul>


### E.1 Creating

In [88]:
t = (1,2,3)
t

(1, 2, 3)

In [89]:
len(t)

3

In [90]:
# mix types
t = ('one', 2, 'Hey!', 3.14, 2)
t

('one', 2, 'Hey!', 3.14, 2)

In [91]:
# index
t[1]

2

In [92]:
# Slicing like lists
t[-1]

2

In [93]:
t[1:3]

(2, 'Hey!')

### E.2 Built-in Methods

In [94]:
# Use .index to enter a value and return the index
t.index(3.14)

3

In [95]:
# Use .count to count the number of times a value appears
t.count(2)

2

Remember it is **immutable**, just like <code>strings</code>.

In [96]:
t[0] = 'new word'

TypeError: 'tuple' object does not support item assignment

In [97]:
t.append('append something')

AttributeError: 'tuple' object has no attribute 'append'

### E.3 Usage

Tuples are not used as often as lists in programming, but are used when immutability is necessary. If in your program you are passing around an object and need to make sure it does not get changed, then a tuple becomes your solution. It provides a convenient source of data integrity.

<hr>

## F. Sets

Unordered collection of **unique** elements


In [98]:
example_set = set()
example_set.add(3)
example_set

{3}

In [99]:
example_set.add('KLM')
example_set

{3, 'KLM'}

In [100]:
# add the same element
example_set.add('KLM')
example_set

{3, 'KLM'}

In [101]:
# Imagine you have a list of repeated items and want to get unique
list_repeat = [1]*3 + [2]*2 + ['KLM']*5 + [1]*2
list_repeat

[1, 1, 1, 2, 2, 'KLM', 'KLM', 'KLM', 'KLM', 'KLM', 1, 1]

In [102]:
set(list_repeat)

{1, 2, 'KLM'}

<hr>

## G. Booleans

In [103]:
a = True
a

True

In [104]:
1 > 2

False

In [105]:
# Placeholder as "None"
b = None
b

In [106]:
b = (1 > 2)
b

False

In [107]:
1 < 2 and 2 < 3

True

In [108]:
1 == 2 or 2 < 3

True

<hr>

## H. Comparison Operators 

Consider <code>a = 3</code> and <code>b=4</code>

<table class="table table-bordered">
<tr>
<th style="width:10%">Operator</th><th style="width:45%">Description</th><th>Example</th>
</tr>
<tr>
<td>==</td>
<td>If the values of two operands are equal, then the condition becomes true.</td>
<td> (a == b) is not true.</td>
</tr>
<tr>
<td>!=</td>
<td>If values of two operands are not equal, then condition becomes true.</td>
<td>(a != b) is true</td>
</tr>
<tr>
<td>&gt;</td>
<td>If the value of left operand is greater than the value of right operand, then condition becomes true.</td>
<td> (a &gt; b) is not true.</td>
</tr>
<tr>
<td>&lt;</td>
<td>If the value of left operand is less than the value of right operand, then condition becomes true.</td>
<td> (a &lt; b) is true.</td>
</tr>
<tr>
<td>&gt;=</td>
<td>If the value of left operand is greater than or equal to the value of right operand, then condition becomes true.</td>
<td> (a &gt;= b) is not true. </td>
</tr>
<tr>
<td>&lt;=</td>
<td>If the value of left operand is less than or equal to the value of right operand, then condition becomes true.</td>
<td> (a &lt;= b) is true. </td>
</tr>
</table>

Know that you can **chain** commands in Python!

And also that *AND* *OR* are also great helpers!

In [109]:
1 < 2 < 3

True

<hr>

## Test!

1. What is the <code>type</code> of the result of the expression 3 + 1.5 + 4?

2. Reverse the string "KLM"

3. Reassign "KLM" on this nested list to say "Air France": <code>example_list = [9,32, [42, 98.9, "KLM"]]</code>

4. Get the value "KLM" from the following dictionaries:
<br><code>d = {'airline': 'KLM'}</code>
<br><code>d = {'flight': {'airline': 'KLM'}}</code>
<br><code>d = {'employee': {'flight': {'flight_number': [889,['KLM']]}}}</code>

5. Can you (by default) sort a dictionary?

6. True or False: Tuple are immutables and lists not.